File: src/lib/scoring/axes.ts
/**
 * The 12-axis scoring rubric. Each axis is scored 0-10 per game.
 * Axes are grouped into 4 "branches" of the skill tree.
 *
 * The order of AXES is load-bearing: score vectors index by position.
 */

export type Branch = "thinking" | "interaction" | "luck" | "experience";

export type AxisKey =
  | "weight"
  | "depth"
  | "density"
  | "interaction"
  | "conflict"
  | "negotiation"
  | "input"
  | "output"
  | "catchup"
  | "theme"
  | "engine"
  | "narrative";

export interface Axis {
  key: AxisKey;
  label: string;
  branch: Branch;
  /** Short experiential implication when this axis is *higher* on game A than on game B. */
  diffExplanation: string;
}

export const AXES: readonly Axis[] = [
  // THINKING — cognition
  { key: "weight",      label: "Weight",      branch: "thinking",    diffExplanation: "more rules to digest, higher learning curve" },
  { key: "depth",       label: "Depth",       branch: "thinking",    diffExplanation: "deeper decision tree, higher skill ceiling" },
  { key: "density",     label: "Density",     branch: "thinking",    diffExplanation: "more meaningful choices per minute, more brain-burn" },

  // INTERACTION — multiplayer effect
  { key: "interaction", label: "Interaction", branch: "interaction", diffExplanation: "players affect each other's plans more, less solitaire" },
  { key: "conflict",    label: "Conflict",    branch: "interaction", diffExplanation: "more direct attacks, friendships tested" },
  { key: "negotiation", label: "Negotiation", branch: "interaction", diffExplanation: "table talk required, diplomacy is mandatory" },

  // LUCK — variance
  { key: "input",       label: "Input",       branch: "luck",        diffExplanation: "more card/setup luck to plan around" },
  { key: "output",      label: "Output",      branch: "luck",        diffExplanation: "more dice/reveal luck, plans can be upended" },
  { key: "catchup",     label: "Catch-up",    branch: "luck",        diffExplanation: "stronger catch-up, no runaway leader" },

  // EXPERIENCE
  { key: "theme",       label: "Theme",       branch: "experience",  diffExplanation: "theme is baked into the mechanics" },
  { key: "engine",      label: "Engine",      branch: "experience",  diffExplanation: "stronger engine-combo arc, explosive late-game turns" },
  { key: "narrative",   label: "Narrative",   branch: "experience",  diffExplanation: "tells a story, consequences persist" },
] as const;

export const AXIS_COUNT = AXES.length;

/** A complete score for a game — one number per axis, in AXES order. */
export type ScoreVector = readonly [
  number, number, number,
  number, number, number,
  number, number, number,
  number, number, number,
];

export const BRANCHES: Record<Branch, { label: string; color: string; tagline: string }> = {
  thinking:    { label: "Thinking",    color: "#58a6ff", tagline: "how much brainwork?" },
  interaction: { label: "Interaction", color: "#f78166", tagline: "how multiplayer is the multiplayer?" },
  luck:        { label: "Luck",        color: "#d4a458", tagline: "where does luck live?" },
  experience:  { label: "Experience",  color: "#8957e5", tagline: "how does it feel?" },
};

export function axisByKey(key: AxisKey): Axis {
  const found = AXES.find((a) => a.key === key);
  if (!found) throw new Error(`Unknown axis: ${key}`);
  return found;
}

export function axisIndex(key: AxisKey): number {
  const idx = AXES.findIndex((a) => a.key === key);
  if (idx < 0) throw new Error(`Unknown axis: ${key}`);
  return idx;
}

File: android/app/src/main/java/com/yournextbg/app/data/scoring/Axes.kt
package com.yournextbg.app.data.scoring

/**
 * 12-axis scoring rubric. Mirrors `src/lib/scoring/axes.ts` 1:1 — same keys,
 * same labels, same order. The order is LOAD-BEARING: score vectors index by
 * position, so a 0..11 reorder corrupts every comparison.
 */

enum class Branch(val key: String, val label: String, val colorHex: String, val tagline: String) {
    THINKING("thinking", "Thinking", "#58a6ff", "how much brainwork?"),
    INTERACTION("interaction", "Interaction", "#f78166", "how multiplayer is the multiplayer?"),
    LUCK("luck", "Luck", "#d4a458", "where does luck live?"),
    EXPERIENCE("experience", "Experience", "#8957e5", "how does it feel?");

    companion object {
        fun byKey(key: String): Branch =
            entries.firstOrNull { it.key == key } ?: error("Unknown branch: $key")
    }
}

enum class AxisKey(val key: String) {
    WEIGHT("weight"),
    DEPTH("depth"),
    DENSITY("density"),
    INTERACTION("interaction"),
    CONFLICT("conflict"),
    NEGOTIATION("negotiation"),
    INPUT("input"),
    OUTPUT("output"),
    CATCHUP("catchup"),
    THEME("theme"),
    ENGINE("engine"),
    NARRATIVE("narrative");

    companion object {
        fun byKey(key: String): AxisKey =
            entries.firstOrNull { it.key == key } ?: error("Unknown axis: $key")
    }
}

data class Axis(
    val key: AxisKey,
    val label: String,
    val branch: Branch,
    val diffExplanation: String,
)

/**
 * The canonical axis list. Position is identity — do NOT reorder.
 *
 * Order checked against `src/lib/scoring/axes.ts` by [AxesTest].
 */
val AXES: List<Axis> = listOf(
    Axis(AxisKey.WEIGHT, "Weight", Branch.THINKING, "more rules to digest, higher learning curve"),
    Axis(AxisKey.DEPTH, "Depth", Branch.THINKING, "deeper decision tree, higher skill ceiling"),
    Axis(AxisKey.DENSITY, "Density", Branch.THINKING, "more meaningful choices per minute, more brain-burn"),

    Axis(AxisKey.INTERACTION, "Interaction", Branch.INTERACTION, "players affect each other's plans more, less solitaire"),
    Axis(AxisKey.CONFLICT, "Conflict", Branch.INTERACTION, "more direct attacks, friendships tested"),
    Axis(AxisKey.NEGOTIATION, "Negotiation", Branch.INTERACTION, "table talk required, diplomacy is mandatory"),

    Axis(AxisKey.INPUT, "Input", Branch.LUCK, "more card/setup luck to plan around"),
    Axis(AxisKey.OUTPUT, "Output", Branch.LUCK, "more dice/reveal luck, plans can be upended"),
    Axis(AxisKey.CATCHUP, "Catch-up", Branch.LUCK, "stronger catch-up, no runaway leader"),

    Axis(AxisKey.THEME, "Theme", Branch.EXPERIENCE, "theme is baked into the mechanics"),
    Axis(AxisKey.ENGINE, "Engine", Branch.EXPERIENCE, "stronger engine-combo arc, explosive late-game turns"),
    Axis(AxisKey.NARRATIVE, "Narrative", Branch.EXPERIENCE, "tells a story, consequences persist"),
)

const val AXIS_COUNT = 12

/**
 * A complete score vector. Must be length 12, value range [0, 10]. The factory
 * validates length so accidental truncation/expansion fails fast.
 */
@JvmInline
value class ScoreVector(val values: DoubleArray) {
    init {
        require(values.size == AXIS_COUNT) {
            "ScoreVector must have $AXIS_COUNT entries, got ${values.size}"
        }
    }

    operator fun get(index: Int): Double = values[index]

    companion object {
        fun of(vararg v: Number): ScoreVector =
            ScoreVector(DoubleArray(v.size) { v[it].toDouble() })

        fun of(v: List<Number>): ScoreVector =
            ScoreVector(DoubleArray(v.size) { v[it].toDouble() })
    }
}

fun axisIndex(key: AxisKey): Int {
    val i = AXES.indexOfFirst { it.key == key }
    require(i >= 0) { "Unknown axis: $key" }
    return i
}

fun axisByKey(key: AxisKey): Axis = AXES[axisIndex(key)]
